### PR TITLE
[`pulse_generator`] Implement `pulse_generator` itself (2/3) 

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
This is part 1 out of 3 into which we split #4073 for ease of reviewing.
See #4073 for all details regarding the overall PR.

**Description of the Change:**
Here we implement the `pulse_generator` gradient method, using the helper methods introduced in the first PR.
The parts missing are the changelog entry and the docstring of `pulse_generator`.

**Related GitHub Issues:**
#4160 (1/3) Helper methods and upgrades to other modules
#4163 (2/3) `pulse_generator` itself
#4166 (3/3) Docstring

### CODE REVIEW

File: pennylane/gradients/pulse_generator_gradient.py [hunk 1]
```diff
@@ -234,3 +250,210 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+        zero_parshapes = iter(zero_parshapes)
+        for i, _g in enumerate(grads):
+            if _g is None:
+                par_shapes = (nonzero_parshape, next(zero_parshapes))
+                # Make zero representative gradient entry, adapting the shape
+                zero_rep = _make_zero_rep(g, single_measure, shot_vector, par_shapes)
+                # Fill in zero-valued gradient entry
+                grads[i] = zero_rep
+
+        grads = [zero_rep if g is None else g for g in grads]
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:413]:
```suggestion
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 2]
```diff
@@ -234,3 +250,209 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+          The type of the Jacobian returned is either a tensor, a tuple or a
+          nested tuple depending on the nesting structure of the original QNode output.
+
+        - If the input is a tape, a tuple containing a
+          list of generated tapes, together with a post-processing
+          function to be applied to the results of the evaluated tapes
+          in order to obtain the Jacobian.
+
+    """
+    argnum = argnum or tape.trainable_params
```
Comment (frederikwilde) [inline pennylane/gradients/pulse_generator_gradient.py:320]: Otherwise `0` gets cast to `False` and higher integers create an error in line 347.
Also, note that `argnums=[]` results in all parameters being differentiated. This is probably not what the user expects.
```suggestion
    if isinstance(argnum, int):
        argnum = [argnum]
    argnum = argnum or tape.trainable_params
```

**Review by frederikwilde — CHANGES_REQUESTED** [review]
Nice work! So much detail in the comments and documentation.
I was able to run most of the code and it works as expected. There is an issue with `reorder_grads()`. I will go over the tests once the code is running.

File: pennylane/gradients/pulse_generator_gradient.py [hunk 3]
```diff
@@ -234,3 +250,209 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+        return qml.math.tensordot(psr_deriv, coeffs, axes=[[0], [0]])
+
+    if single_measure and not shot_vector:
+        # single measurement and single shot entry
+        return _parshift_and_contract_single(results, qml.math.stack(coeffs))
+    if single_measure or not shot_vector:
+        # single measurement or single shot entry, but not both
+        return tuple(
+            _parshift_and_contract_single(r, qml.math.stack(coeffs)) for r in zip(*results)
+        )
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:282]: I like this! I'll go one step further and change the input argument of this helper method to `single_shot_entry` :)

File: pennylane/gradients/pulse_generator_gradient.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:320]: Mhm. Good point. I think this behaviour exists throughout the module. I like the fix for integers,but would go with the possibly unexpected behaviour for `argnum=[]`  for now. We can change this behaviour (with a deprecation cycle) in another PR :)

Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:320]: Opened #4213 .

File: pennylane/gradients/pulse_generator_gradient.py [hunk 4]
```diff
@@ -234,3 +250,209 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+            if _g is None:
+                par_shapes = (nonzero_parshape, next(zero_parshapes))
+                # Make zero representative gradient entry, adapting the shape
+                zero_rep = _make_zero_rep(g, single_measure, shot_vector, par_shapes)
+                # Fill in zero-valued gradient entry
+                grads[i] = zero_rep
+
+
+        # Reformat the flat list of gradients into an output that fits the original tape specs.
+        return reorder_grads(grads, tape_specs)
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:415]: Thanks. Taking a look!

**Comment by frederikwilde** [comment]
Also, when I run this example with a `QNode` I get an unexpected output:
```Python
import jax
import jax.numpy as jnp
jax.config.update('jax_enable_x64', True)

import pennylane as qml
from pennylane.pulse import ParametrizedEvolution, ParametrizedHamiltonian
from pennylane.gradients.pulse_generator_gradient import pulse_generator

coeff1 = lambda p, t: p[0] * jnp.sin(t * p[1]) + p[2]
obs1 = qml.PauliX(0)
param1 = jnp.array([1., 2., 3.], dtype=jnp.float64)

coeff2 = lambda p, t: jnp.cos(p * t)
obs2 = qml.PauliX(1)
param2 = jnp.array(.5)

ham = ParametrizedHamiltonian([coeff1, coeff2], [obs1, obs2])
dev = qml.device('default.qubit.jax', wires=2)

@qml.qnode(dev)
def qnode(param1, param2):
    ParametrizedEvolution(ham, (param1, param2), t=1.)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

grad_fn = pulse_generator(qnode)
grad_fn(param1, param2)
```
This gives
```Python
Array([[[0.43605857, 0.        , 0.        ],
        [0.        , 0.43605857, 0.        ],
        [0.        , 0.        , 0.43605857]],

       [[0.0619607 , 0.        , 0.        ],
        [0.        , 0.0619607 , 0.        ],
        [0.        , 0.        , 0.0619607 ]],

       [[0.61583807, 0.        , 0.        ],
        [0.        , 0.61583807, 0.        ],
        [0.        , 0.        , 0.61583807]]], dtype=float64)
```
instead of
```Python
(Array([0.43605857, 0.0619607 , 0.61583807], dtype=float64),
 Array(0.12961439, dtype=float64))
```
which is what one gets with a `QuantumTape`.

**Comment by dwierichs** [comment]
Hi @frederikwilde,
>Also, when I run this example with a QNode I get an unexpected output:

Regarding this, we will not have an immediate bug fix, I'm afraid. When applying gradient transforms to a `QNode` directly, `qml.transforms.classical_jacobian` is used to map between the Jacobian of the tape and the one of the QNode. However, `classical_jacobian` does not support non-scalar tape parameters yet.
I'll open an issue for this, but I don't think it's realistic to fix as a side task within this PR.
I added a test for this level of differentiation (`gradient_transform` on `QNode`) for scalar tape parameters, as well as tests for non-scalar and mixed parameters that are marked with `pytest.xfail` for now.

File: pennylane/gradients/pulse_generator_gradient.py [hunk 5]
```diff
@@ -234,3 +250,210 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+        for r in zip(*results)
+    )
+
+
+def _expval_pulse_generator(tape, argnum, shots, atol):
+    """Compute the pulse generator parameter-shift rule for a quantum circuit that returns expectation
+    values of observables.
+
+    Args:
+        tape (`~.QuantumTape`): Quantum circuit to be differentiated with the pulse generator
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/gradients/pulse_generator_gradient.py:288]: Or maybe not, if it currently can't be?

File: pennylane/gradients/pulse_generator_gradient.py [hunk 6]
```diff
@@ -234,3 +250,210 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+    # Extract some specifications about the original tape, required for output formatting
+    num_measurements = len(tape.measurements)
+    single_measure = num_measurements == 1
+    num_params = len(tape.trainable_params)
+    shot_vector = isinstance(shots, Sequence)
+    tape_specs = (single_measure, num_params, num_measurements, shot_vector, shots)
+
+    def processing_fn(results):
+        """Post-process the results of the parameter-shifted tapes of the pulse
+        generator parameter-shift rule into the gradient."""
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:368]: Indeed a bit convoluted sentence 😅 Maybe just "Post-process the results of the parameter-shifted tapes into the gradient"

File: tests/gradients/core/test_pulse_generator_gradient.py [hunk 7]
```diff
@@ -693,3 +703,621 @@ def _int_sin(p, t):
...
+            return qml.expval(Z(0))
+
+        params = [jnp.array(0.4)]
+        with qml.Tracker(dev) as tracker:
+            grad = jax.jacobian(circuit)(params)
+        p = params[0] * T
+        exp_grad = -2 * jnp.sin(2 * p) * T
+        assert qml.math.allclose(grad, exp_grad)
+        assert tracker.totals["executions"] == 1 + 2  # one forward pass, two shifted tapes
+
```
Comment (frederikwilde) [inline tests/gradients/core/test_pulse_generator_gradient.py:1155]: I would also test QNodes with multiple arguments here. This way I actually found an error in the `processing_fn`. As far as I can tell, this test
```suggestion

@pytest.mark.parametrize('argnums', [[0, 1], 0, 1])
    def test_simple_qnode_expval_multiple_params(self, argnums):
        """Test that a simple qnode with two parameters
        can be differentiated with pulse_generator and `argnums` works as expected."""
        import jax
        import jax.numpy as jnp

        jax.config.update("jax_enable_x64", True)
        dev = qml.device("default.qubit.jax", wires=1)
        T = 0.2
        ham1 = qml.pulse.constant * Y(0)
        ham2 = qml.pulse.constant * Y(0)

        @qml.qnode(dev, interface="jax", diff_method=pulse_generator)
        def circuit(param1, param2):
            qml.evolve(ham1)(param1, T)
            qml.evolve(ham2)(param2, T)
            return qml.expval(Z(0))

        param1, param2 = [jnp.array(0.4)], [jnp.array(0.5)]
        with qml.Tracker(dev) as tracker:
            grad = jax.jacobian(circuit, argnums=argnums)(param1, param2)
        p = (param1[0] + param2[0]) * T
        exp_grad = -2 * jnp.sin(2 * p) * T
        if argnums == [0, 1]:
            for g in grad:
                assert qml.math.allclose(g, exp_grad)
                assert tracker.totals["executions"] == 1 + 4  # one forward pass, four shifted tapes

        else:
            assert qml.math.allclose(grad, exp_grad)
            assert tracker.totals["executions"] == 1 + 2  # one forward pass, two shifted tapes

```
should pass, but somehow `nonzero_parshape` never gets defined when `argnums=1` or `argnums=[1]` because [this line](https://github.com/PennyLaneAI/pennylane/blob/hybrid-pulse-diff-2b/pennylane/gradients/pulse_generator_gradient.py#L392) is never reached.

**Review by frederikwilde — CHANGES_REQUESTED** [review]
Really thorough work with the tests! 👍 
I only see two issues with `argnum` and `argnums` for JAX. For the first one I unresolved a previous code suggestion that would fix this. Or should this be fixed in a different way? For the second issue with `argnums=1` for QNodes, I don't exactly understand what's going wrong, or is this related to the previous issue with QNodes? In this case both parameters are scalars, though.

As a remark: I would always advocate for setting PRNG seeds and in the case of NumPy using `np.random.default_rng(seed)`. But there are also a benefits to having a different seed each time the test runs, so I think it's fine either way.

File: pennylane/gradients/pulse_generator_gradient.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:313]: Thanks, I wanted to incorporate this before but somehow missed to do it in the end. Included the above now.

File: pennylane/gradients/pulse_generator_gradient.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:368]: Fair point. :laughing: 
I changed it to @Qottmann's suggestion, including a mention of `pulse_generator` to make it reference where it comes from :)

File: pennylane/gradients/pulse_generator_gradient.py [hunk 8]
```diff
@@ -234,3 +250,210 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
...
+        grads = []
+        zero_parshapes = []
+        # Iterate over gradient_data, which is equivalent to the trainable parameters in argnum
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:371]: Much nicer. Yes, that's indeed what I meant :)

File: tests/gradients/core/test_pulse_generator_gradient.py [hunk 9]
```diff
@@ -693,3 +703,621 @@ def _int_sin(p, t):
...
+        params = pnp.array([0.5, 0.5, 0.5], requires_grad=True)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.Rot(*params, wires=0)
+            qml.probs([2, 3])
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+
+        res_pulse_gen, _tapes = grad_fn(tape, dev)
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/gradients/core/test_pulse_generator_gradient.py:888]: - The operation will have a zero gradient, as determined by `gradient_analysis_and_validation`, because it acts on a wire that has no impact on the measurement at all.
- `_assert_has_jax` only checks whether `JAX` is installed.
- The check for operations being a `ParametrizedEvolution` is made further down the line, in `_generate_tapes_and_coeffs`. This makes sense because a non-pules operation in a tape is fine as long as we don't want to tackle it with the `pulse_generator` method, or it does not affect the measurements as in the test example :)

File: tests/gradients/core/test_pulse_generator_gradient.py (hunk 7 above)
Comment (dwierichs) [reply tests/gradients/core/test_pulse_generator_gradient.py:1155]: Excellent catch! And thanks so much for providing a full-fletched test :heart_eyes: 
Indeed there was a small bug in the code, [using the wrong index](https://github.com/PennyLaneAI/pennylane/blob/d73a696ffc5c4b782dac5aaa395af3b7f8fc0f75/pennylane/gradients/pulse_generator_gradient.py#L340) to decide whether to create a gradient entry or a zero dummy.

**Comment by dwierichs** [comment]
Just to summarize:

> For the first one I unresolved a previous code suggestion that would fix this.

suggestion applied.

> For the second issue with argnums=1 for QNodes, I don't exactly understand what's going wrong, or is this related to the previous issue with QNodes? In this case both parameters are scalars, though.

This was a bug, fixed it and included the test you suggested.

> I would always advocate for setting PRNG seeds and in the case of NumPy using np.random.default_rng(seed). But there are also a benefits to having a different seed each time the test runs, so I think it's fine either way.

Good catch! I presume that you refer to tests with shots, is that correct? I'll include seeds for those tests.

File: tests/gradients/core/test_pulse_generator_gradient.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (frederikwilde) [reply tests/gradients/core/test_pulse_generator_gradient.py]: Oh, that all makes sense. Thanks for clarifying!

File: pennylane/gradients/pulse_generator_gradient.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/gradients/pulse_generator_gradient.py:288]: The "If the input is a QNode," blurb below kind of throws me then, because it makes it sound like an option - but I guess this is how the other ones do it too 😅

**Review by frederikwilde — APPROVED** [review]
The new tests look great to me! All good to go 💯
